### PR TITLE
Remove unnecessary pure annotations

### PR DESCRIPTION
**Pull request by Keno**
These should now be unnecessary. Our effect modeling of Core.sizeof
has improved sufficiently, that it should be able to know right away
that these functions are pure without the additional annotation.

### CODE REVIEW

**Review by Sacha0 — APPROVED** [review]
Nifty! :)

**Comment by Keno** [comment]
These were added by @JeffBezanson in https://github.com/JuliaLang/julia/commit/0b80774b9f0a517f12714e30808915e99fd916ff, which was https://github.com/JuliaLang/julia/pull/32437 where I now realize I already said that this shouldn't be `@pure`. At least I'm consistent :)

**Comment by KristofferC** [comment]
Nice to get rid of the last correct usages of `@pure` (according to the docs).

**Comment by vtjnash** [comment]
We explicitly block constant-prop of String, which I think is what made this necessary

**Comment by Keno** [comment]
Ok, yes, I see (after this PR): 
```
julia> function bar()
       return Base.sizeof("foo")
       end
bar (generic function with 1 method)

julia> bar()
3

julia> @code_typed bar()
CodeInfo(
1 ─ %1 = Core.sizeof::typeof(Core.sizeof)
│   %2 = (%1)("foo")::Int64
└──      return %2
) => Int64
```
Fortunately, I'm about to fix that too, so we'll just have to hold this PR for a little bit.

**Comment by Keno** [comment]
> about to fix that too

The "about to fix that" is #43852, so this can go in once that's merged.

**Pull request closed without merging**
